### PR TITLE
feat: Update GitHub functions

### PR DESCRIPTION
**Pull request by eric-nguyen-cs**
### What
- Replace `PyGithub` with `Githubkit` for async HTTP requests
- Fix Github export, by using `AsyncTransactionContext` in `BackgroundTasks` (this also allows for `controllers` code reuse)
- Export from the correct SHA when creating a new PR (used new `project_controllers` and `project_models`)
- Add a new commit when exporting an exported taxonomy (instead of raising an `GtihubBranchAlreadyExists` error)

### Fixes bug(s)
Fixes https://github.com/openfoodfacts/taxonomy-editor/issues/349

### CODE REVIEW

File: backend/editor/api.py [hunk 1]
```diff
@@ -103,27 +105,6 @@ def check_single(id):
...
 
 
-def file_cleanup(filepath):
-    """
-    Helper function to delete a taxonomy file from local storage
-    """
-    try:
-        os.remove(filepath)
-    except FileNotFoundError:
-        log.warn(f"Taxonomy file {filepath} not found for deletion")
```
Comment (alexgarel) [inline backend/editor/api.py:113]: good idea to move it.

**Review by alexgarel — APPROVED** [review]
Great PR.

Globally ok, if we can keep tests, it would be better !

File: backend/pyproject.toml [hunk 2]
```diff
@@ -48,10 +44,10 @@ readme = "README.md"
 [tool.poetry.dependencies]
 python = "^3.11"
 fastapi = { extras = ["all"], version = "^0.104.1" }
+githubkit = { version = "^0.11.0a4", allow-prereleases = true }
```
Comment (perierc) [inline backend/pyproject.toml:47]: why allow prereleases here specifically? I'm curious

File: backend/editor/github_functions.py [hunk 3]
```diff
@@ -13,12 +21,22 @@ class GithubOperations:
...
-        self.repo = self.init_driver_and_repo()
+        self.repo_info = self.get_repo_info()
+        self.connection = self.init_connection()
 
-    def init_driver_and_repo(self):
+    def get_repo_info(self) -> tuple[str, str]:
+        repo_uri = settings.repo_uri
+        if not repo_uri:
+            raise HTTPException(
+                status_code=400, detail="repo_uri is not set. Please add your access token in .env"
```
Comment (perierc) [inline backend/editor/github_functions.py:34]:
```suggestion
                status_code=400, detail="repo_uri is not set. Please add the repo URI in .env"
```

File: backend/editor/github_functions.py [hunk 4]
```diff
@@ -33,47 +51,68 @@ def init_driver_and_repo(self):
             raise HTTPException(
                 status_code=400, detail="repo_uri is not set. Please add your access token in .env"
```
Comment (perierc) [inline backend/editor/github_functions.py:52]:
```suggestion
                status_code=400, detail="repo_uri is not set. Please add the repo URI in .env"
```

File: backend/editor/entries.py [hunk 5]
```diff
@@ -26,6 +29,8 @@
     TransactionCtx,
     get_current_transaction,
 )
+from .models.project_models import ProjectCreate, ProjectEdit, ProjectStatus
+from .utils import file_cleanup  # Normalizing tags
```
Comment (perierc) [inline backend/editor/entries.py:33]:
```suggestion
from .utils import file_cleanup
```

**Review by perierc — COMMENTED** [review]
Good work, happy with the new models and controller, a few nitpicks

File: backend/pyproject.toml (hunk 2 above)
Comment (eric-nguyen-cs) [reply backend/pyproject.toml:47]: If I remember correctly it was to get the latest version of `githubkit` that was compatible for `Pydantic V2`

File: backend/editor/github_functions.py [hunk 6]
```diff
@@ -33,47 +51,68 @@ def init_driver_and_repo(self):
...
+            if e.response.status_code == 404:
+                return None
+            raise e
+        return result.parsed_data
 
-    def list_all_branches(self):
+    async def get_file_sha(self) -> str:
         """
-        List of all current branches in the "openfoodfacts-server" repo
+        Get the contents of a file in the "openfoodfacts-server" repo
```
[the remark this answers is not in the document]
Comment (eric-nguyen-cs) [reply backend/editor/github_functions.py]: Yes thanks!

File: backend/tests/test_api.py [hunk 7]
```diff
@@ -41,21 +33,7 @@ def test_ping(client):
...
-    # which downloads the taxonomy file from a Github URL
-    mocker.patch("editor.api.TaxonomyGraph.import_from_github", return_value=True)
-
-    response = client.post(
-        "/test/testing_branch/import",
-        json={"description": "test_description"},
-    )
-
-    assert response.status_code == 200
-    assert response.json() is True
```
[the remark this answers is not in the document]
Comment (eric-nguyen-cs) [reply backend/tests/test_api.py:55]: Not really sure what value this tests brings, as the only difference with the `/upload` endpoint is that we have to access Github, which should be mocked

File: backend/tests/test_api.py [hunk 8]
```diff
@@ -67,20 +45,6 @@ def test_upload_taxonomy(client, github_mock):
...
-
-    with open("tests/data/test.txt", "rb") as f:
-        response = client.post(
-            "/test_taxonomy_2/test_branch/upload",
-            files={"file": f},
-            data={"description": "test_description"},
-        )
-
-    assert response.status_code == 409
-    assert response.json() == {"detail": "branch_name: Branch name should be unique!"}
```
[the remark this answers is not in the document]
Comment (eric-nguyen-cs) [reply backend/tests/test_api.py:81]: Same here, I believe this checks if the mock function works correctly and if we check for duplicate branches by using the `get_all_branches` endpoint of `Github` (that is not the case anymore, we check if the branch we want to create already exists or not). 
My philosophy on tests follows quite closely what this engineer describes in this video: https://www.youtube.com/watch?v=MbU-PKukdMw
If that's ok, I'll merge the PR and if you want to add the test back, I'll do that in another PR

File: backend/tests/test_api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexgarel) [reply backend/tests/test_api.py:55]: I hear the first part of the video you mentioned in [other comment](https://github.com/openfoodfacts/taxonomy-editor/pull/368#discussion_r1474901611). I'm ok with what is said in the video. I also value more integration test than unit test if functions are just assembling stuff.

But here, this is no unit test, this is integration test, that's why I think it's useful.

**Pull request merged**